spectator-ext-jvm/src/main/java17/com/netflix/spectator/jvm/JavaFlightRecorder.java

```diff
@@ -0,0 +1,107 @@
+package com.netflix.spectator.jvm;
+
+import com.netflix.spectator.api.Registry;
+import jdk.jfr.EventSettings;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordingStream;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+
+/**
+ * Helpers supporting continuous monitoring with Java Flight Recorder.
+ */
+public class JavaFlightRecorder {
+
+  private static final String PREFIX = "jdk.";
+  private static final String CPULoad = PREFIX + "CPULoad";
+  private static final String VirtualThreadPinned = PREFIX + "VirtualThreadPinned";
+  private static final String VirtualThreadSubmitFailed = PREFIX + "VirtualThreadSubmitFailed";
+  private static final String ZAllocationStall = PREFIX + "ZAllocationStall";
+  private static final String ZYoungGarbageCollection = PREFIX + "ZYoungGarbageCollection";
+  private static final String ZOldGarbageCollection = PREFIX + "ZOldGarbageCollection";
+
+  private JavaFlightRecorder() {
+  }
+
+  /**
+   * Collect low-overhead Java Flight Recorder events, using the provided
+   * {@link Executor} to execute a single task to collect events.
+   *
+   * @param registry the registry
+   * @param executor the executor to execute the task for streaming events
+   * @return an {@link AutoCloseable} allowing the underlying
+   * {@link jdk.jfr.consumer.EventStream} to be closed
+   */
+  public static AutoCloseable monitorDefaultEvents(Registry registry, Executor executor) {
+    Objects.requireNonNull(registry);
+    Objects.requireNonNull(executor);
+    RecordingStream rs = new RecordingStream();
+    collectCpuEvents(registry, rs);
+    collectVirtualThreadEvents(registry, rs);
+    collectZgcEvents(registry, rs);
+    executor.execute(rs::start);
+    return rs::close;
+  }
+
+  private static void collectCpuEvents(Registry registry, RecordingStream rs) {
+    consume(CPULoad, rs, event -> {
+      String pid = "unknown";
+      try {
+        pid = String.valueOf(ProcessHandle.current().pid());
+      } catch (UnsupportedOperationException ignore) {
+      }
+      registry.gauge("jvm.cpuLoad", "type", "jvmUser", "pid", pid)
+        .set(event.getFloat("jvmUser"));
+      registry.gauge("jvm.cpuLoad", "type", "jvmSystem", "pid", pid)
+        .set(event.getFloat("jvmSystem"));
+      registry.gauge("jvm.cpuLoad", "type", "machineTotal", "pid", pid)
+        .set(event.getFloat("machineTotal"));
```

Comment:
Would this overlap with the contribution from user/system? That would make aggregation confusing, e.g. if a user doesn't break down by type.

Reply:
Yeah, it does. My thinking about the the CPU load was being able to reason about per-process CPU load (thus the `pid` for differentiating when you have more than one VM per-host), for multitenant JVM use cases and accounting for primary process versus sidecars. But if you're concerned about the duplication, I can totally drop this for now.

Reply:
In this comment it is more a question of the modeling. Users should be able to incrementally drill into the data. So if the user just selects `name,jvm.cpuLoad,:eq,:sum`, then they would be aggregating over all the types. If `machineTotal` overlaps with the other types, then they get a confusing result. But yeah, if we just drop the CPU stuff we can ignore that.